syntax = "proto2";

import "NetworkMessage.proto";

package SimpleChat;

enum CommandType {
    MOTD = 0;
    KICK = 1;
    MUTE = 2;
    UNMUTE = 3;
    AUTH = 4;
    UNKNOWN = 5;
}

message ChatTarget {
    required string user_name = 1;
    optional int32 user_id = 2;
}

message ChatMessage {
    required string text = 1;

    optional ChatTarget from = 2;
    optional ChatTarget target = 3;
}

message ChatCommand {
    required CommandType type = 2;
    repeated string arguments = 3;
}

message ChatroomChange {
    optional string motd = 1;
}

message GenericChatResponse {
    required bool success = 1;
    optional string message = 2;
}

extend NetworkMessage {
  optional ChatMessage chatMessage = 20;
  optional ChatCommand chatCommand = 21;

  optional ChatroomChange chatroomChange = 22;
  optional GenericChatResponse genericChatResponse = 23;
}